onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -287,6 +287,86 @@ static inline void RegisterExecutionProvider(InferenceSession* sess, onnxruntime
   OrtPybindThrowIfError(sess->RegisterExecutionProvider(std::move(p)));
 }
 
+#ifdef USE_CUDA
+static inline void RegisterCudaExecutionProviderWithCache(InferenceSession* sess, const ProviderOptionsMap& provider_options_map) {
+  if (auto* cuda_provider_info = TryGetProviderInfo_CUDA()) {
+    const auto it = provider_options_map.find(kCudaExecutionProvider);
+    CUDAExecutionProviderInfo info{};
+    if (it != provider_options_map.end())
+      cuda_provider_info->CUDAExecutionProviderInfo__FromProviderOptions(it->second, info);
+    else {
+      info.device_id = cuda_device_id;
+      info.gpu_mem_limit = gpu_mem_limit;
+      info.arena_extend_strategy = arena_extend_strategy;
+      info.cudnn_conv_algo_search = cudnn_conv_algo_search;
+      info.do_copy_in_default_stream = do_copy_in_default_stream;
+      info.external_allocator_info = external_allocator_info;
+    }
+
+    // This variable is never initialized because the APIs by which is it should be initialized are deprecated, however they still
+    // exist are are in-use. Neverthless, it is used to return CUDAAllocator, hence we must try to initialize it here if we can
+    // since FromProviderOptions might contain external CUDA allocator.
+    external_allocator_info = info.external_allocator_info;
+
+    static std::unordered_map<OrtDevice::DeviceId, std::shared_ptr<IExecutionProvider>> cuda_eps;
+
+    auto device_id = info.device_id;
+    if (cuda_eps.find(device_id) == cuda_eps.end()) {
+      auto cuda_ep_factory = cuda_provider_info->CreateExecutionProviderFactory(info);
+      cuda_eps[device_id] = std::move(cuda_ep_factory->CreateProvider());
+    }
+    OrtPybindThrowIfError(sess->RegisterExecutionProvider(cuda_eps[device_id]));
```

Comment:
API question: What should happen in the options in `info` are different from the ones used to construct the cached EP? 

This was a common source of complaints in TensorFlow 1.x, where the configuration of the *first* `Session` determined some global options (e.g. the sizes of some global threadpools), and then some parts of the configuration for subsequent sessions would be ignored.

I guess this won't be a problem for the `ORTModule` use case, since it doesn't expose options for setting these provider options to the user. But it could be an issue for Python users of the `InferenceSession` API.

Reply:
good point... I think I can change it to take the entire CUDAExecutionProviderInfo struct as the key for the cache.